src/lib/y2partitioner/actions/add_bcache.rb

```diff
@@ -0,0 +1,69 @@
+# encoding: utf-8
+
+# Copyright (c) [2018] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "y2partitioner/dialogs/bcache"
+require "y2partitioner/device_graphs"
+
+Yast.import "Popup"
+
+module Y2Partitioner
+  module Actions
+    class AddBcache
+      def run
+        result = {}
+        dialog = Dialogs::Bcache.run(device_graph, result)
+        return :cancel if dialog != :next
+
+        create_device(result)
+        :finish
+      end
+
+    private
+
+      def create_device(result)
+        backing = device_graph.blk_devices.find{ |d| d.name == result[:backing] }
+        raise "Invalid result #{result.inspect}. Backing not found." unless backing
+
+        caching = device_graph.blk_devices.find{ |d| d.name == result[:caching] }
+        raise "Invalid result #{result.inspect}. Backing not found." unless backing
+
+        bcache_nums = device_graph.bcaches.map(&:basename).map { |n| n[/bcache(\d+)/, 1].to_i }
+        new_num = (bcache_nums.max || -1) + 1
+        bcache = backing.create_bcache("/dev/bcache#{new_num}")
+
+        bcache_cset = device_graph.bcache_csets.find { |d| d.blk_devices.include?(caching) }
+        if !bcache_cset
+          if caching.respond_to?(:partition_table)
+            caching.remove_descendants
+          end
+          bcache_cset = caching.create_bcache_cset
+        end
+
+        bcache.attach_bcache_cset(bcache_cset)
+      end
```

Comment:
Again, all this logic about reusing or creating cache sets is something I would expect in the `Y2Storage` classes (if not even in the library).

Reply:
on which level of abstraction? Like method in Bcache like `bcache.add_caching(blk_device)` or in BcacheCset like `BcacheCset.find_or_create(blk_device)` ?

Reply:
At first glance, both look ok to me